### PR TITLE
Cosmos: Fixes around array projection

### LINKED ISSUE #33797: Cosmos: Support projecting out (primitive) collections
We need to generally think about how this is handled, not just in Cosmos; in relational, ToList() is handled very specifically in RelationalProjectionBindingEV - not sure if e.g. ToArray() works.

Also, think about whether ToList() should be handled purely as a Select() phenomenon (i.e. in ProjectionBindingEV). In Cosmos, it probably makes sense for ToList() to be generally translatable to [`ARRAY()`](https://learn.microsoft.com/en-us/azure/cosmos-db/nosql/query/subquery#array-expression), i.e. as a way to convert the results of a subquery to a collection scalar. This could also make sense in relational (e.g. array_agg on PG). On the other hand, ToList() seems to be useful mostly/only in projections, since one can simply e.g. compare the results of a subquery directly to some array (`Where(b => b.Ints.Where(...) == b.Ints2` - this applies ARRAY() as well).

### PR DESCRIPTION
**Pull request by roji**
Along with #34008, this should complete the collection projection story (though I still have a few things left to investigate).

I'm not super happy with the way things work, but this is more or less aligned with the relational behavior, and mostly works... I can see potential for cleanup and improvements in the future, but I've already spent enough time on it...

Closes #33797

### CODE REVIEW

File: test/EFCore.Cosmos.FunctionalTests/Query/PrimitiveCollectionsQueryCosmosTest.cs [hunk 1]
```diff
@@ -1751,16 +1788,18 @@ public override async Task Project_multiple_collections(bool async)
                 """
 SELECT VALUE
 {
-    "Ints" : c["Ints"],
     "c" : ARRAY(
+        SELECT VALUE i
+        FROM i IN c["Ints"]),
```
Comment (AndriySvyryd) [inline test/EFCore.Cosmos.FunctionalTests/Query/PrimitiveCollectionsQueryCosmosTest.cs:1793]: Wasn't `c["Ints"]` a better translation?

**Review by AndriySvyryd — APPROVED** [review]

File: test/EFCore.Cosmos.FunctionalTests/Query/PrimitiveCollectionsQueryCosmosTest.cs (hunk 1 above)
Comment (roji) [reply test/EFCore.Cosmos.FunctionalTests/Query/PrimitiveCollectionsQueryCosmosTest.cs:1793]: Yeah, it was - this is the result of composing ToList() on top of a bare array (ToList() is one of the major things enabled by this PR)... This is actually quite complicated, since the CLR type of the array (`int[]`) can't just be cast (in the shaper) to a `List<int>`.

FWIW the same thing happens in relational - if you look at the SQL Server SQL baseline for this same test you'll see `OUTER APPLY OPENJSON([p].[Ints]) AS [i]`, although we could have just projected `[p].[Ints]` directly and parsed the JSON... It's indeed an unfortunate edge-case - opened #34081 to track improving it across both relational and Cosmos.
